src/engine.ml

```diff
@@ -175,7 +204,13 @@ let incoming_tls tls data =
                    Fmt.(option ~none:(unit "no") Cstruct.hexdump_pp) out
                    Fmt.(option ~none:(unit "no") Cstruct.hexdump_pp) d);
       Error (`Tls e)
-    | `Ok tls' -> Ok (tls', out, d)
+    | `Ok tls' ->
+      if expect_no_reply then
+        match out with
+        | None -> Ok (tls', out, d)
+        | Some _ -> Error (`Msg "expected no TLS reply")
+      else
+        Ok (tls', out, d)
```

Comment:
Would it maybe be cleaner to implement `no_expect_reply` as a wrapper that doesn't return `out`?
Like
```ocaml
let incoming_tls_expect_no_reply tls data = incoming_tls tls data >>= function
  | Ok (tls', None, d) -> Ok (tls', d)
  | Ok (_, Some _, _) -> Error (`Msg "expected no TLS reply")
  | Error _ as e -> e
```
Ignoring the `out` after passing the optional argument at each call site looks a bit fragile.

Reply:
that's a good suggestion, done in caef48b